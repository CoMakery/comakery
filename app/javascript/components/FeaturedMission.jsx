import React from 'react'
import PropTypes from 'prop-types'
import InterestsController from '../controllers/interests_controller'

export default class FeaturedMission extends React.Component {
  constructor(props) {
    super(props)

    this.addInterest = this.addInterest.bind(this)
    this.removeInterest = this.removeInterest.bind(this)
    this.state = {
      name       : props.name,
      description: props.description,
      imageUrl   : props.imageUrl,
      projects   : props.projects
    }
  }

  addInterest(projectId) {
    new InterestsController().follow(projectId).then(response => {
      if (response.status === 200) {
        let newProjects = Array.from(this.state.projects)
        const index = newProjects.findIndex(project => project.id === projectId)
        newProjects[index].interested = true
        this.setState({projects: newProjects})
      } if (response.status === 401) {
        window.location = '/accounts/new'
      } else {
        throw Error(response.text())
      }
    })
  }

  removeInterest(projectId) {
    new InterestsController().unfollow(projectId).then(response => {
      if (response.status === 200) {
        let newProjects = Array.from(this.state.projects)
        const index = newProjects.findIndex(project => project.id === projectId)
        newProjects[index].interested = false
        this.setState({projects: newProjects})
      } else if (response.status === 401) {
        window.location = '/accounts/new'
      } else {
        throw Error(response.text())
      }
    })
  }

  render() {
    const { name, description, imageUrl, projects } = this.state
    const { missionUrl } = this.props

    return (
      <div className={`featured-mission ${this.props.float === 'right' ? 'featured-mission--right' : ''}`}>
        <a href={missionUrl} className={`featured-mission__image ${this.props.float === 'right' ? 'featured-mission__image--right' : ''}`}>
          <img src={imageUrl} />
          <span>view mission</span>
        </a>
        <div className="featured-mission__detail">
          <div className="featured-mission__name">
            <a href={missionUrl}>{name}</a>
          </div>
          <div className="featured-mission__description">{description}</div>
          {projects.map(project => <div key={project.id} className="featured-mission__project">
            <a href={`/projects/${project.id}`} className="featured-mission__project__title">{project.title}</a>
            {project.interested &&
              <div
                className="featured-mission__project__interest featured-mission__project__interest"
                onClick={() => { this.removeInterest(project.id) }}
              >
                Unfollow
              </div>
            }
            {!project.interested &&
              <div
                className="featured-mission__project__interest"
                onClick={() => { this.addInterest(project.id) }}
              >
                Follow
              </div>
            }
          </div>)}
          <a href={`/projects/new${this.props.id ? `?mission_id=${this.props.id}` : ''}`} className="featured-mission__create-project">Create New Project</a>
        </div>
      </div>
    )
  }
}

FeaturedMission.propTypes = {
  float      : PropTypes.string,
  id         : PropTypes.number,
  name       : PropTypes.string,
  description: PropTypes.string,
  imageUrl   : PropTypes.string,
  projects   : PropTypes.array,
  csrfToken  : PropTypes.string,
  missionUrl : PropTypes.string
}

FeaturedMission.defaultProps = {
  float      : 'left',
  id         : undefined,
  name       : '',
  description: '',
  imageUrl   : '',
  projects   : [],
  csrfToken  : '00',
  missionUrl : ''
}
