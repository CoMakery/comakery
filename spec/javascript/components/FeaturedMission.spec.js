import React from 'react'
import { mount } from 'enzyme'
import FeaturedMission from 'components/FeaturedMission'

describe('FeaturedMission', () => {
  beforeEach(() => {
    fetch.resetMocks()
  })

  it('renders correctly without props', () => {
    const wrapper = mount(<FeaturedMission />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.featured-mission')).toBe(true)
    expect(wrapper.exists('.featured-mission__image')).toBe(true)
    expect(wrapper.exists('.featured-mission__detail')).toBe(true)
    expect(wrapper.exists('.featured-mission__name')).toBe(true)
    expect(wrapper.exists('.featured-mission__description')).toBe(true)
  })

  it('renders correctly with props', () => {
    const mission = {
      float      : 'right',
      name       : 'Test Mission Name',
      symbol     : 'BTC',
      description: 'This is test mission',
      imageUrl   : '/logo.png',
      projects   : [
        { id: 1, title: 'Project 1', interested: true },
        { id: 2, title: 'Project 2', interested: false }
      ],
      csrfToken: '00'
    }

    const wrapper = mount(<FeaturedMission {...mission} />)

    expect(wrapper.exists('.featured-mission--right')).toBe(true)
    expect(wrapper.find('.featured-mission__image').find('img').prop('src')).toEqual('/logo.png')
    expect(wrapper.find('.featured-mission__name').text()).toEqual('Test Mission Name')
    expect(wrapper.find('.featured-mission__description').text()).toEqual('This is test mission')
    expect(wrapper.find('.featured-mission__project').first().find('.featured-mission__project__title').text()).toEqual('Project 1')
    expect(wrapper.find('.featured-mission__project').first().find('.featured-mission__project__interest--sent').text()).toEqual('Request Sent')
    expect(wrapper.find('.featured-mission__project').last().find('.featured-mission__project__title').text()).toEqual('Project 2')
    expect(wrapper.find('.featured-mission__project').last().find('.featured-mission__project__interest').text()).toEqual('Iâ€™m interested')
  })

  it('updates project-interest', () => {
    const mission = {
      float      : 'right',
      name       : 'Test Mission Name',
      symbol     : 'BTC',
      description: 'This is test mission',
      imageUrl   : '/logo.png',
      projects   : [
        { id: 1, title: 'Project 1', interested: true },
        { id: 2, title: 'Project 2', interested: false }
      ],
      csrfToken: '00'
    }

    const wrapper = mount(<FeaturedMission {...mission} />)
    wrapper.setState({
      projects: [
        { id: 1, title: 'Project 1', interested: true },
        { id: 2, title: 'Project 2', interested: true }
      ]
    })

    wrapper.update()
    expect(wrapper.find('.featured-mission__project').first().find('.featured-mission__project__title').text()).toEqual('Project 1')
    expect(wrapper.find('.featured-mission__project').first().find('.featured-mission__project__interest--sent').text()).toEqual('Request Sent')
    expect(wrapper.find('.featured-mission__project').last().find('.featured-mission__project__title').text()).toEqual('Project 2')
    expect(wrapper.find('.featured-mission__project').last().find('.featured-mission__project__interest--sent').text()).toEqual('Request Sent')
  })
})
