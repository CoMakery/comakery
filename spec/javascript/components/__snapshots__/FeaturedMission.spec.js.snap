// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturedMission renders correctly without props 1`] = `
<FeaturedMission
  csrfToken="00"
  description=""
  float="left"
  imageUrl=""
  name=""
  projects={Array []}
  symbol=""
>
  <div
    className="featured-mission "
  >
    <div
      className="featured-mission__image "
    >
      <img
        src=""
      />
    </div>
    <div
      className="featured-mission__detail"
    >
      <div
        className="featured-mission__name"
      >
         
      </div>
      <div
        className="featured-mission__description"
      />
      <a
        className="featured-mission__create-project"
        href="/projects/new"
      >
        Create New Project
      </a>
    </div>
  </div>
</FeaturedMission>
`;
